bigquery_etl/backfill/utils.py

```diff
@@ -1,51 +1,223 @@
+        sql_dir, qualified_table_name
     )
 
-    for file in backfill_files:
-        backfills = Backfill.entries_from_file(file)
+    if backfill_file.exists():
+        backfills = Backfill.entries_from_file(backfill_file, status)
         if backfills:
-            backfills_dict[file].extend(backfills)
+            backfills_dict[qualified_table_name] = backfills
 
     return backfills_dict
+
+
+def get_qualified_table_name_to_entries_map_by_project(
+    sql_dir, project_id, status=None
+) -> defaultdict[str, List[Backfill]]:
+    """Return backfill entries from project."""
+    backfills_dict_all: defaultdict = defaultdict(list)
+
+    search_path = Path(sql_dir) / project_id
+    backfill_files = Path(search_path).rglob(BACKFILL_FILE)
+
+    for backfill_file in backfill_files:
+        project, dataset, table = extract_from_query_path(backfill_file)
+        qualified_table_name = f"{project}.{dataset}.{table}"
+
+        backfills_dict = get_qualified_table_name_to_entries_dict(
+            sql_dir, qualified_table_name, status
+        )
+
+        backfills_dict_all.update(backfills_dict)
+
+    return backfills_dict_all
+
+
+def get_backfill_file_from_qualified_table_name(sql_dir, qualified_table_name) -> Path:
+    """Return backfill file from qualified table name."""
+    project, dataset, table = qualified_table_name_matching(qualified_table_name)
+    path = Path(sql_dir)
+    query_path = path / project / dataset / table
+    backfill_file = query_path / BACKFILL_FILE
+
+    return backfill_file
+
+
+def get_backfill_staging_qualified_table_name(qualified_table_name, entry_date) -> str:
+    """Return full table name where processed backfills are stored."""
+    project, dataset, table = qualified_table_name_matching(qualified_table_name)
+
+    backfill_table_id = f"{table}_{entry_date}".replace("-", "_")
+
+    return f"{BACKFILL_DESTINATION_PROJECT}.{BACKFILL_DESTINATION_DATASET}.{backfill_table_id}"
+
+
+def validate_metadata_workgroups(sql_dir, qualified_table_name) -> bool:
+    """
+    Return True if metadata workgroup is mozilla-confidential or None.
+
+    The backfill staging dataset currently only support backfilling tables for workgroup:mozilla-confidential.
+    When workgroup is None, the default (workgroup:mozilla-confidential) will be applied.
+    """
+    valid_workgroup = ["workgroup:mozilla-confidential"]
+
+    project, dataset, table = qualified_table_name_matching(qualified_table_name)
+    dataset_path = Path(sql_dir) / project / dataset
+
+    query_files = Path(dataset_path).rglob("*.sql")
+
+    for query_file in query_files:
+        try:
+            # check table level metadata
+            table_metadata_path = query_file.parent / METADATA_FILE
+            table_metadata = Metadata.from_file(table_metadata_path)
+            table_workgroup_access = table_metadata.workgroup_access
+
+            if table_workgroup_access is not None:
```

Comment:
Add comment or function to explain reasoning (empty list) for line 121-127 and line 134-140